### PR TITLE
#35 Use taxon and name IDs in species/match

### PR DESCRIPTION
**Pull request by timrobertson100**
Note: All previous work you've seen is merged into a single commit in this PR.

### CODE REVIEW

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    if (backboneUsageKey != null) {
+      LOG.debug("Backbone key[{}] found using scientificNameID[{}]. No further taxon IDs will be used", backboneUsageKey, scientificNameID);
+      issues.add(TAXON_MATCH_TAXON_ID_IGNORED);
+      issues.add(TAXON_MATCH_TAXON_CONCEPT_ID_IGNORED);
+    } else {
+      LOG.debug("Attempting to locate a backbone taxon for taxonID[{}]", taxonID);
+      backboneUsageKey = lookupBackboneKey(taxonID, scientificName, issues, TAXON_MATCH_TAXON_ID_IGNORED, TAXON_ID_NOT_FOUND);
+    }
+
+    if (backboneUsageKey != null) {
```
Comment (MattBlissett) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:59]: Won't this still be true if it was found on line 48 with the scientificNameId? i.e. should this block be in the else block ending on line 57.

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    String scientificNameID = identification.getScientificNameID();
+    String taxonID = identification.getTaxonID();
+    String taxonConceptID = identification.getTaxonConceptID();
+
+    LOG.debug("Attempting to locate a backbone taxon for scientificNameID[{}]", scientificNameID);
+    backboneUsageKey = lookupBackboneKey(scientificNameID, scientificName, issues, TAXON_MATCH_SCIENTIFIC_NAME_ID_IGNORED, SCIENTIFIC_NAME_ID_NOT_FOUND);
+
+    if (backboneUsageKey != null) {
+      LOG.debug("Backbone key[{}] found using scientificNameID[{}]. No further taxon IDs will be used", backboneUsageKey, scientificNameID);
+      issues.add(TAXON_MATCH_TAXON_ID_IGNORED);
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:52]: should these be flagged only if there are values present?

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      issues.add(TAXON_MATCH_TAXON_CONCEPT_ID_IGNORED);
+    } else {
+      LOG.debug("Attempting to locate a backbone taxon for taxonID[{}]", taxonID);
+      backboneUsageKey = lookupBackboneKey(taxonID, scientificName, issues, TAXON_MATCH_TAXON_ID_IGNORED, TAXON_ID_NOT_FOUND);
+    }
+
+    if (backboneUsageKey != null) {
+      LOG.debug("Backbone key[{}] found using taxonID[{}]. No further taxon IDs will be used", backboneUsageKey, taxonID);
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:60]: Will this not also be entered if the key was found via the scientificNameID?

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      issues.add(TAXON_MATCH_TAXON_CONCEPT_ID_IGNORED);
+    } else {
+      LOG.debug("Attempting to locate a backbone taxon for taxonConceptID[{}]", taxonConceptID);
+      backboneUsageKey = lookupBackboneKey(taxonConceptID, scientificName, issues, TAXON_MATCH_TAXON_CONCEPT_ID_IGNORED, TAXON_CONCEPT_ID_NOT_FOUND);
+    }
+
+    LOG.debug("Backbone key was not found using any of the name or taxa IDs");
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:67]: same here. You should always reach this whether found or not

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        // Short circuit: if the checklist is already pointing to the backbone we don't need any further lookup
+        if (checklistKey.equalsIgnoreCase(NUB_DATASET_KEY.toString())) {
+          try {
+            backboneUsageKey = Integer.parseInt(adjustedID);
+          } catch (NumberFormatException e) {
+            issues.add(notFoundFlag);
+            LOG.debug("Adjusted ID[{}] is no backbone key integer", adjustedID);
+          }
+
+        } else {
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:89]: when GBIF ids are found no flags are raised

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    try {
+      ParsedName recordName = PARSER.parse(name, null, null);
+      ParsedName foundName = PARSER.parse(match.getScientificName(), null, null);
+      boolean equal = recordName!=null && foundName!=null &&
+          StringUtils.equals(recordName.canonicalName(), foundName.canonicalName());
+      if (!equal) {
+        LOG.debug("Canonical name[{}] found using ID[{}] does not match canonical name[{}] on record, after parsing", foundName.canonicalName(), id, recordName.canonicalName());
+        issues.add(SCIENTIFIC_NAME_AND_ID_INCONSISTENT); // we will still use the found concept though
+      }
+    } catch(Exception e) {
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:134]: Maybe catch UnparsableNameException explicitly as these are rather likely to occur and handle only as warning?

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+  /**
+   * Add a flag only if the provided name and match response can be parsed and the canonical names found to differ.
+   * This lenient in behaviour as it is only a hint to publishers so should not fail.
+   */
+  static void compareAndFlagDifferingNames(String id, String name, Set<OccurrenceIssue> issues, NameUsageSearchResponse.Result match) {
+    try {
+      ParsedName recordName = PARSER.parse(name, null, null);
+      ParsedName foundName = PARSER.parse(match.getScientificName(), null, null);
+      boolean equal = recordName!=null && foundName!=null &&
```
Comment (mdoering) [inline kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:128]: if one of the parsed names would be null they would be unequal and you would get an NPE in the below log line.
If I am correct the parser would never return null parsed names but instead throw an UnparsableNameException, so these null checks here can be removed.

One of the first steps in the name parser does this:
>     if (StringUtils.isBlank(scientificName)) {
      throw new UnparsableNameException(NameType.NO_NAME, scientificName);

**Review by mdoering — CHANGES_REQUESTED** [review]

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 2 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:52]: Good point. Yes

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 3 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:60]: No - the IF above means it was found

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 4 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:67]: good point - bad log line

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 5 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:89]: Good point, I'll add

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 6 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:134]: Will do

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 7 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:128]: good catch. will fix

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 3 above)
Comment (MattBlissett) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:60]: The logic works, but the log message will be inaccurate if the key was found via scientificNameID.

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 5 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java:89]: Correction: I think it's working correctly with regard to this.
If a backbone key is found in scientificNameID, then the logical will add ignored flags on the taxonID and taxonConceptID

**Comment by timrobertson100** [comment]
@mdoering @MattBlissett - thank you very much
Please see changes, noting I changed the order of the IFs for readability 

Markus - I think the flags are added correctly if a backbone ID is used directly. It works the same if a correct WoRMS ID is found and used.

File: kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java (hunk 3 above)
Comment (timrobertson100) [reply kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java]: Fixed in next commit

**Review by mdoering — APPROVED** [review]
That looks all good and solves all the issues I've flagged before

**Comment by timrobertson100** [comment]
Thank you @MattBlissett @mdoering 

I am going to squash the commits into one, rewriting history and then self-merge

**Pull request merged**